### ISSUE TITLE
Making a package for Primitive Permutation Groups Library

### ISSUE THREAD

**Issue by alex-konovalov** [gap-system/gap#791]
This library (http://www.gap-system.org/Datalib/prim.html) should become a proper package, loadable via `LoadPackage`. The library is in the separate `prim` directory, and the documentation is already in GAPDoc format at http://www.gap-system.org/Manuals/doc/ref/chap50.html#X7AE00EA7791F2574.

For the beginning, I suggest to make it needed package for GAP for backwards compatibility (so GAP will not start without it). Packages that rely on it (CTblLib and IRREDSOL, maybe more) would then be able to include it in their dependencies. Eventually we will put it into the list of packages loaded by default but not needed to start GAP. 

Questions:
- License: GPL v2?
- Authors: http://www.gap-system.org/Datalib/prim.html says "Colva Roney-Dougal and many others", and gives Colva's and @hulpke details for contacts. Who should be the authors of the package?
- package name: "prim" or "PrimitiveGroups" or "PrimitiveGroupsLibrary"? My current choice is "PrimitiveGroups" as the name of the directory (note that `LoadPackage("prim");`) will still work as long as there is no other package started with "prim").

I am happy to do the conversion which is pretty straightforward. I am setting GAP 4.9.0 as a milestone and assigning it to myself.

**Comment by alex-konovalov** [comment]
This should now incorporate #818 which extends it for degrees up to 4095

**Comment by dimpase** [comment]
What kind of backwards compatibility are you talking about? GAP 4.8.\* starts and runs without this library just fine:
```
% gap
 ┌───────┐   GAP 4.8.3, 19-Mar-2016, build of 2016-05-16 05:03:46 (UTC)
 │  GAP  │   http://www.gap-system.org
 └───────┘   Architecture: x86_64-unknown-linux-gnu-gcc-default64
 Libs used:  gmp, readline
 Loading the library and packages ...
 Packages:   GAPDoc 1.5.1
 Try '?help' for help. See also  '?copyright' and  '?authors'
gap> 
```
Please do not make things needed, if they are not...

**Comment by alex-konovalov** [comment]
@dimpase: missing `trans` causes errors while loading Browse package, and missing `prim` - errors for CTblLib and IRREDSOL (not surprising, since package authors of course assume that these are parts of the core system). They will need time to adapt their packages and add new "TransitiveGroups" and "PrimitiveGroups" packages to their dependencies.

**Comment by dimpase** [comment]
Why don't you instead just add these dependencies to these packages?
Or send them (trivial) patches to apply...

Instead you somehow prefer to intentionally break GAP.

**Comment by markuspf** [comment]
On Wed, Jun 15, 2016 at 05:01:48AM -0700, Dima Pasechnik wrote:
> Why don't you instead just add these dependencies to these packages?
> Or send them (trivial) patches to apply...
> 
> Instead you somehow prefer to intentionally break GAP.

I think the way Alex suggests how to go about making these components packages
is absolutely sensible. We make a change, and don't break backwards
compatibility for a while.

In addition we can send those patches, or suggest to people that they should
update their packages.

**Comment by dimpase** [comment]
@markuspf : Alex suggests to break backwards compatibility for people who do not need trans and prim, by co-opting the latter into "GAP core".

**Comment by alex-konovalov** [comment]
@dimpase I don't understand how this could be "co-opting". They became independent packages, so taken out of the GAP core where they are now, and nothing will change for people who do not need trans and prim at all.

**Comment by dimpase** [comment]
And, by the way, it  looks more like a bug in Browse that it needs trans to load, for I cannot imagine Browse *using* trans functionality for its inner workings; similarly CTblLib probably does not really need prim.

**Comment by dimpase** [comment]
@alex-konovalov : for your own version of "core" you might be right. But trans and prim are data libraries, no more than that.

**Comment by fingolfin** [comment]
While I disagree with @dimpase rhetorics, I also don't see why we should make the new packages "needed" -- wouldn't it be sufficient to have them autoload by default? I.e. always load them if they are present, but if they are not, just don't?

Eventually, they would then be moved to the regular default list of autoloaded packages, which user's can overwrite. We'd do that *after* all relevant packages had time to adapt (i.e. to add explicit dependencies to the new trans / prim packages).

It's important to do these things incremental, to allow a smooth transition.

**Comment by fingolfin** [comment]
However, let me point out that right now, we *do* consider trans and prim to be parts of GAP core. Yes, you can remove them and GAP still works. And if you wantede to, you could also remove support for, say, permutation groups from GAP, and GAP could still work (it would be a lot more effort to achieve that, of course). So what? Doing it is not supported. Moving them into packages is what will make it "supported", so I'd think you should be happy about this...

That said, even if we marked those new packages as "needed", then people who want to hack GAP by removing these GAP core parts can simply hack GAP a little more, and edit the bits where we mark the packages as needed.

But again, I don't think we need to mark them "needed".

**Comment by alex-konovalov** [comment]
@dimpase : neither of this is a bug.

- Browse provides `BrowseTransitiveGroupsInfo` which shows the contents of the transitive groups library: http://www.gap-system.org/Manuals/pkg/Browse/doc/chapA.html#X832846F486391385

- `ctbllib/gap4/ctdbattr.g` uses `PRIMRANGE`, `AllTransitiveGroups` and `AllPrimitiveGroups` to setup mappings between the character table library and these group libraries

- IRREDSOL package relies on `DegreeOfMatrixGroup` defined in `prim//irredsol.gd`.

**Comment by alex-konovalov** [comment]
@fingolfin : I have actually though about the same (making them autoloaded by default) prompted by @dimpase inquiries. Here is why I've declined that idea: if one starts GAP with `-A` option, then default packages (defined in `package.gi` as `[ "autpgrp", "alnuth", "crisp", "ctbllib", "factint", "fga",               "irredsol", "laguna", "polenta", "polycyclic", "resclasses", "sophus", "tomlib" ],` and to be extended with `PrimitiveGroups`, `TransitiveGroups` and maybe more) will not be loaded. Now try to load IRREDSOL, Browse, or CTblLib, and you will still get an error.

**Comment by dimpase** [comment]
@alex-konovalov : saying that `Browse` must depend on `trans` is akin to saying that  `Firefox` must be bundled with contents of `www.gap-system.org`. 

And I never said anything about IRREDSOL.

**Comment by dimpase** [comment]
@fingolfin : given that your humble servant belong to *people who want to hack on GAP*, also by virtue of being part of the same project called OpenDreamKit, I feel unhappy when other members of the same project saddle me with perfectly avoidable work. So it's OK to saddle me and other Sage (or Debian) contributors with extra work, and it is *Verboten* as soon as GAP package authors are concerned.  

And it is while everybody agrees that ultimately this *needed* hack is not needed in the long run.
You still may not like this rhetoric, but why not fix it now and not some years later, if at all?

**Comment by bh11** [comment]
As far as IRREDSOL is concerned, it uses just four definitions from prim/irredsol.gd, namely
```
DeclareSynonymAttr ("DegreeOfMatrixGroup", DimensionOfMatrixGroup);
DeclareProperty("IsPrimitiveMatrixGroup", IsMatrixGroup);
DeclareAttribute("MinimalBlockDimension", IsMatrixGroup);
DeclareSynonymAttr ("IsLinearlyPrimitive", IsPrimitiveMatrixGroup);
```
Moving these four declarations from `prim/` to `lib/`would resolve the depencency. (The dependency merely exists to make the primitive soluble group databases in prim/ and IRREDSOL more compatible from a user's point of view.) Getting rid of `IsLinearlyPrimitive` in IRREDSOL wouldn't be a problem at all, but I'd like to keep DegreeOfMatrixGroup  as a synonym for `DimensionOfMatrixGroup`. 

Or maybe drop the irredsol part from prim/ altogether, given that the same functionality (but for a vastly larger range of groups) is in IRREDSOL?

Anyway, this should be easy to resolve quickly.

**Comment by alex-konovalov** [comment]
@dimpase Browse package does nothing wrong here, and seeing a dependency that does not directly relate to the package title is not a justification for suspecting a bug. Browse legitimately assumes that the transitive groups library is present as the part of the (current) GAP core system, and just uses it. 

The plan for a "one step at a time" transition is as follows: 
- Transitive library converted to a proper GAP package
- Browse adds it to dependencies (as suggested, with loading appropriate part of the code only when it's present?)
- CTblLib adds it to dependencies (as required?)
- After the new releases of both packages are picked up for the redistribution, the transitive group package may be removed from needed and moved to loaded by default.

This is smooth and gradual, and doing it faster will break regression tests and increase a risk of overlooking more serious problems.

**Comment by markuspf** [comment]
Can I ask politely for any kind of personal chat to be taken off this tracker,
please?

**Comment by dimpase** [comment]
@alex-konovalov : it's all fine as long as your "one step at a time" transition is complete within beta releases before 4.9, and not "eventually", as you suggest...

**Comment by alex-konovalov** [comment]
To clarify, assigning this to GAP 4.9 milestone means making the 1st step (Transitive library converted to a proper GAP package) before the beta release. Next steps depend on two packages, and they should not hold the first public release of GAP 4.9 in case their updates will be delayed.

**Comment by alex-konovalov** [comment]
@bh11 I think moving these four declarations from `prim/` to `lib/` would be helpful. If not, it could happen that there is some user code also relying on them, and that code will became broken too.

I don't know about dropping the `irredsol` part from prim/ altogether. is that same code, or different implementation of the same functionality? In the latter case, it could be still useful to have it for cross-checks. Also, that may be done after converting it to a proper package, too.

**Comment by alex-konovalov** [comment]
Let me summarise where we are now with the initial questions. Nobody suggested anything else, so the new package will be under GPL v2 as core GAP system and will have the name "PrimitiveGroups". But I still don't have all details about authors: http://www.gap-system.org/Datalib/prim.html says "Colva Roney-Dougal and many others", and gives Colva's and @hulpke details for contacts. Who else should be the author of the PrimitiveGroups package?

**Comment by dimpase** [comment]
Max sees offence and rudeness just everywhere, it seems. 

> Date: Wed, 15 Jun 2016 16:39:15 +0200

**Comment by markuspf** [comment]
@dimpase it is *completely* inappropriate to quote personal emails in public. I ask you to reconsider this move very carefully. I will have to close this discussion if this does not stop. Please take these discussions somewhere else.

**Comment by frankluebeck** [comment]
The plan described by @alex-konovalov is the correct procedure:
  1. move primitive groups into a package (indeed very desirable)
  2. make this new package "needed" for GAP (to give other packages some time to introduce a new dependency or to change their code in some other way taking into account that the primitive groups are not necessarily available)
  3. when other packages are adjusted remove the primitive groups package from the list of "needed" packages and add it to the default list of packages loaded at startup (no need to hurry)

Step 2. just keeps the situation as it is now, nothing becomes worse, and this is the only way to guarantee that the primitive groups are available. 

If some distribution hacks GAP (probably incompletely) to start  without primitive groups then it must use a different (much easier) hack, but that sounds acceptable for me.

**Comment by bh11** [comment]
@alex-konovalov The irredsol part is different from IRREDSOL (it has been used to cross check IRREDSOL, of course). 

BTW., the irredsol part of prim is based upon work by Mark Short (Sho92 in the GAP bibliography), not sure if he wrote the GAP code and/or should be mentioned as an author.

**Comment by dimpase** [comment]
@frankluebeck : "make the new package needed" changes the setup for GAP packaging teams, if it is implemented `no needed packages found, bailing out` (as suggested by Alex).

- making the package needed is not necessary for the distributions provided by GAP, as it would be bundled in, anyway;  
- on the other hand, this check will most likely be removed by 3rd party GAP distributors.

So having such step 2 implemented as in a stable release makes no sense. We have had long discussions on closely related topics in #776 and #788; why are we having it again now?

**Comment by ChrisJefferson** [comment]
@dimpase : GAP and it's packages are not a monolithic project. If we made the package not a needed package, then it would be hard to have a working browse. We would have to change browse, GAP core, and introduce the new package all on the same day, and none of them would be compatible with the packages from the day before.

By doing things in phases (change to required package, then fix browse) we remove the need to do everything at once in a "big bang". As @alex-konovalov said, if we extract to a non-needed package then that breaks the current release of Browse (and other packages). You may prefer if GAP and it's packages were treated as a single monolithic entity, but that is not currently the case.

**Comment by alex-konovalov** [comment]
The package is now available at http://gap-packages.github.io/primgrp/. See  https://github.com/gap-packages/primgrp/issues/5 for the plan of removing `prim` from the core system in several steps (hope may be done in a couple of days).

**Comment by alex-konovalov** [comment]
This may be closed after merging #1650.

**Comment by alex-konovalov** [comment]
The package PrimGrp now resides here: 
  * repository: https://github.com/gap-packages/primgrp
  * website: http://gap-packages.github.io/primgrp/
and `prim` directory has been removed in #1650.
